rafthttp/snapshot.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rafthttp
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/pkg/ioutil"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+// snapshotAutoReleaseTimeout is the timeout before the kept snapshot is
+// auto released.
+//
+// The timeout needs to be greater than the maximal time that may take
+// before the kept snapshot is retrieved.
+// Before a snapshot is retrieved, its raft snapshot message needs to
+// be sent out, and snapHandler waits remote member requesting
+// snapshot data. The process needs to take two maximal RTT, which is 10s.
+// So we set 15s as the timeout.
+const snapshotAutoReleaseTimeout = 15 * time.Second
```

Comment:
I use const because this is simpler. I know that a short snapshotAutoReleaseTimeout could help etcd to attempt snapshot send fast, but 15s is good enough considering snapshot send is a rare case.